### PR TITLE
BZ1270708 Allow using different encoding based on User-Agent

### PR DESCRIPTION
**Pull request by spyrkob**
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1312064

Undertow issue: https://issues.jboss.org/browse/UNDERTOW-595
Undertow commit: https://github.com/undertow-io/undertow/commit/98fe959a4f4cde414c36b268734e22421f170460

### CODE REVIEW

File: src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java [hunk 1]
```diff
@@ -33,15 +40,35 @@
  */
 public abstract class BasicAuthenticator extends Authenticator {
 
+    public static final Charset DEFAULT_CHARSET = Charset.forName("UTF-8");
     protected String realm;
 
+    public Map<Pattern, Charset> browserCharsetMap;
```
Comment (wolfc) [inline src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java]: How is this configured in EAP6?
Why is it public?

File: src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java [hunk 2]
```diff
@@ -33,15 +40,35 @@
...
+     * on the browser that issued it.
+     *
+     * @param realm The HTTP Basic authentication realm
+     * @param defaultCharset charset that should be used to decode credentials if the user agent is not listed in browserCharsetMap
+     * @param browserCharsetMap map indexed by Patterns representing User-Agent strings with the charset as values
+     */
+    public BasicAuthenticator (String realm, Charset defaultCharset, Map<Pattern, Charset> browserCharsetMap) {
         this.realm = realm;
+        this.defaultCharset = defaultCharset;
+        this.browserCharsetMap = new HashMap<Pattern, Charset>(browserCharsetMap);
```
Comment (wolfc) [inline src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java]: Should this be a copy or un-modifiable instance?

Comment (spyrkob) [inline src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java]: I assumed a copy would be enough for a private field - is it better to wrap it in an unmodifiableMap?

File: src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java (hunk 1 above)
Comment (spyrkob) [inline src/main/java/org/jboss/com/sun/net/httpserver/BasicAuthenticator.java]: example configuration in EAP6: https://github.com/jbossas/jboss-eap/pull/2727/files

browserCharsetMap should be private - will fix.

**Pull request merged**